### ISSUE TITLE
clangd causes headers to be flagged as not used

### ISSUE THREAD

**Issue by cjdb** [clangd/clangd#1683]
**Please describe the problem.**
https://reviews.llvm.org/D152686 seems to have broken clangd for a project of mine. Most of my headers are now flagged as not being directly used even though they're the root header for the symbol. 

I've managed to bisect it to this commit, but don't really know how to reduce it for a repro. Should I just try and recreate one of the headers in a new project with my config file? Should I also consider my compile_commands.json?

Interestingly, if I ask clangd to take me to the definition, it will move to that file.

**Logs**

[clangd.txt](https://github.com/clangd/clangd/files/11872187/clangd.txt)

**System information**

Any version of clangd post-D152686.

Editor/LSP plugin: VS Code

Operating systems: Arch Linux, macOS SSHing to Debian

**Comment by kadircet** [comment]
Hi @cjdb, your compile flags are also likely to be relevant, but usually having some header & source code should be enough to demonstrate a small reproducer for include-cleaner purposes.

I was looking into the logs attached and there seems to be patterns like:
foo.h:
```
#pragma once
namespace std { int declval(int); }
```
foo.cc:
```
#include "foo.h"
void foo() { std::declval(5); }
```

Does this relate to the patterns you're facing the issues with?
If so in this particular case, we actually use a custom mapping for standard library symbols, with the assumption that extending std namespace is undefined behaviour [\[1\]](https://en.cppreference.com/w/cpp/language/extending_std).

So if your project involves such patterns the best thing is to disable analysis for the whole project or certain headers. See https://clangd.llvm.org/guides/include-cleaner#adjusting-behavior for some configuration knobs.

**Comment by cjdb** [comment]
Thanks for identifying the root problem so quickly!

> If so in this particular case, we actually use a custom mapping for standard library symbols, with the assumption that extending std namespace is undefined behaviour [1]. 

Ah, that explains why it's affecting this project and not others. That's unfortunate, since this project is a standard library implementation (the `std::declval` you see is the one from `<utility>`). I suspect that it's mapping to libstdc++ at the moment, which isn't ideal.

Would it be possible for clangd to use normal mappings when `nostdinc++` is passed as an option? That would make it possible to use this feature with the broadest set of projects.

**Comment by HighCommander4** [comment]
> Would it be possible for clangd to use normal mappings when `nostdinc++` is passed as an option?

My sense is some projects use `nostdinc++` in situations where conceptually they _are_ using the standard library (i.e. it's not a project like libc++ where you're working on the standard library itself), but they're providing the standard library include paths manually rather than having clangd determine them (e.g. because the `--query-driver` mechanism doesn't work for some reason).

So, we may want a separate config knob for this use case.

**Comment by cjdb** [comment]
Perhaps there should be an option in `.clangd` along the lines of `StdlibDevelopment: True`?

**Comment by vogelsgesang** [comment]
> some projects use nostdinc++ in situations where conceptually they are using the standard library (i.e. it's not a project like libc++ where you're working on the standard library itself), but they're providing the standard library include paths manually 

Yes, the project which I am working on at work also fits this description: We are using the standard library, but we want to overwrite the include paths and hence use `nostdinc++`

**Comment by osirisgothra** [comment]
Since this is kinda an ongoing thing I think at this point, I can offer my advice for dealing with this as it can be overwhelming to users who are unfamiliar with clangd. Now, if you have IWYU, you can use the in-comment macros after each include to prevent this from happening:

`#include "secretly-used.h" // IWYU pragma: keep`

If you use QtCreator, and are not familiar with clangd and just want rid of the error you can disable it for one or more files, or for your entire project (or globally) by doing the following:

1. Click the projects (Wrench Icon) on the left sidebar.
2. For each project there will be an inner sidebar, click the active project you wish to edit.
3. On that subsidebar, under the "Project Settings" heading, go down until you see "Clangd" and click on that
4. Uncheck "use global settings"
5. The very last line will say "User specific settings go [here]"  Click on the link under the "[here]" text.
6. An editor window will open, you will want to add the following lines to disable unused include checking for that project:

If you dont use QtCreator, you likely already know where your user settings yaml file is, just edit it and join in at this point or open the yaml file manually if you use some other IDE:

 ```
 Diagnostics:
    UnusedIncludes: None

```
7. Alternatively, you can instead just restrict one or more filename patterns to not scan, if you still want the missing header thing to work on your other source files:

```
  Diagnostics:
     Includes:
         IgnoreHeader:  ["<regex pattern>",".*\.h"]
```

 Note that these ignored patterns are matched against the filename string at the suffix. So using ".*reg" is the same as "reg" and will match a file named "file.reg" or "thisisareg" but not "regfile" or "fileregfile". To get prefix matching you need to use "^TEXT.*" the "^" indicates you want to match the beginning of the item. All Items are assumed to end with '$' which means end of string, so it does not need to be included nor should it appear anywhere unescaped in the string.

Hope that helps anyone who lands here.
